src/software/ai/evaluation/calc_best_shot.h

```diff
@@ -31,6 +32,25 @@ namespace Evaluation
                                            const std::vector<Point> &obstacles,
                                            double radius);
 
+    /**
+     * Finds the best shot on the given goal, and returns the best target to shoot at and
+     * the largest open angle interval for the shot (this is the total angle between the
+     * obstacles on either side of the shot vector).
+     *
+     * @param goal_post_neg The goal post of the net with the negative y-coordinate
+     * @param goal_post_pos The goal post of the net with the positive y-coordinate
+     * @param p The point that the shot will be taken from
+     * @param robot_obstacles The locations of any robots on the field that may obstruct
+     * the shot. These are treated as circular obstacles, and are usually used to
+     * represent robots on the field.
+     *
+     * @return the best target to shoot at and the largest open angle interval for the
+     * shot (this is the total angle between the obstacles on either side of the shot
+     * vector). If no shot is possible, returns `std::nullopt`
+     */
+    std::optional<Shot> calcBestShotOnGoal(const Point &goal_post_neg,
+                                           const Point &goal_post_pos, const Point &p,
+                                           const std::vector<Robot> &obstacles);
```

Comment:
We have 3 calcBestShotOnGoal and the signatures are only subtly different so it's a lot of work for a code reader to quickly know which calcBestShotOnGoal is being called. Can we give them different names (WithObstacles, WithTeams, etc)?

Reply:
Hmm, where do we draw the line about when we should not override a function and should instead give it a different name?

Reply:
in discussion offline on a general approach but it's resolved for this PR